### PR TITLE
AI Bat Enemy

### LINKED ISSUE #62: Advanced AI for bat enemy

### PR DESCRIPTION
**Pull request by ethanr04**
## What's been done
- I have integrated a new Neural Network-based AI for the Bat enemy, using Unity's ML Agents package.
- Overhaul to BatPlayground scene, now functions as a training scene with 20 simultaneous simulations/areas. Each area will include one or two bats, a player and an obstacle. These will be randomly generated each episode to prevent overfitting.
- Created a BatAgent.cs script that uses Reinforcement Learning to navigate complex geometry, flank the player, and perform attacks. Used a heurstic function to reward the bats for certain things, such as being on the opposite sides of the player and successfully hitting them.
- Multi-Agent enemy behaviour, with implicit coordination as a result of training 
- Made use of Ray Perception Sensors so the bats can navigate the environment
- Produced a few different models:
  - ```FlankingDuoBatBrain.onnx``` is the main, most up-to-date one. Bats were training both together and independently to stop the 'cautious' behaviour. 
  - ```CautiousDuoBatBrain.onnx``` works for the most part, but when the bats become seperated, the bat still targetting the player will keep its distance and rarely attack. May be useful to show intelligence, i.e. waiting for backup.
  -  ```SelfishBatBrain.onnx``` a fairly basic model that was only trained with a single bat in a simple non-dynamic environment. Fairly similar to standard enemy logic (but probably worse!)
- Integrated new enemy superclass for cleaner and more readable code. Only thing I had to change was removing the 'abstract' part from the superclass, as the bats already inherit from the Agent class. To get it to work, the bats have access to the base from a script.
 
https://github.com/user-attachments/assets/19f3177b-39f5-4292-a30b-f5a493f2d101

## How to use
- Bats have been made as prefabs, couple of different variants for sleeping bats, and independent (selfish) bats
- Drag into the scene, assign ```Player``` and ```Other Bat``` fields under the agent script in the inspector. If there is only 1 bat, it will function fine if you leave the ```Other Bat``` field empty. 
- Ensure ```Training Mode``` field is unchecked
- Under Behaviour Parameters, drag the desired model into the ```Model``` field.
- Press play, and they should function as intended. Bear in mind the AI only kicks in during the 'Chase' state.

## How to train (optional)
- Ensure that mlagents toolkit for Python is installed.
- Training parameters are specified in the config folder, in ```BatConfig```, and can be changed to modify fields such as the max number of iterations
- Activate your python environment by running ```.\venv\Scripts\activate```
- Change directory to the Assets folder in the Unity project. 
- Run ```mlagents-learn [path_to_config_file] --run-id=[name_of_folder_for_brain_to_go]```. Include the ```--resume``` flag at the end in order to continue from a pre-existing brain.
- It should now be running and waiting for you to run the BatPlayground scene.
- When you are done, or it finishes, you can CTRL-C and it should save the brain to the specified folder

https://github.com/user-attachments/assets/473c2cd9-e08e-4ab7-ab1d-87349198c555

### CODE REVIEW

**Review by megan-pa — DISMISSED** [review]
lgtm :)

**Review by qc23151 — DISMISSED** [review]
LGTM

**Review by OliverTchum — DISMISSED** [review]
LGTM :D

**Review by OliverTchum — APPROVED** [review]
LGTM :D

**Review by megan-pa — APPROVED** [review]
yay
